akka-sample-cqrs-scala/src/main/scala/sample/cqrs/ShoppingCart.scala

```diff
@@ -0,0 +1,191 @@
+
+    def updateItem(itemId: String, quantity: Int): State = {
+      quantity match {
+        case 0 => copy(items = items - itemId)
+        case _ => copy(items = items + (itemId -> quantity))
+      }
+    }
+
+    def removeItem(itemId: String): State =
+      copy(items = items - itemId)
+
+    def checkout(): State = copy(checkedOut = true)
+
+    def toSummary: Summary =
+      Summary(items, checkedOut)
+  }
+  object State {
+    val empty = State(Map.empty, checkedOut = false)
+  }
+
+  /**
+   * This interface defines all the commands that the ShoppingCart persistent actor supports.
+   */
+  sealed trait Command extends CborSerializable
+
+  /**
+   * A command to add an item to the cart.
+   *
+   * It can reply with `Confirmation`, which is sent back to the caller when
+   * all the events emitted by this command are successfully persisted.
+   */
+  final case class AddItem(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command
+
+  /**
+   * A command to remove an item from the cart.
+   */
+  final case class RemoveItem(itemId: String, replyTo: ActorRef[Confirmation]) extends Command
+
+  /**
+   * A command to adjust the quantity of an item in the cart.
+   */
+  final case class AdjustItemQuantity(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command
+
+  /**
+   * A command to checkout the shopping cart.
+   */
+  final case class Checkout(replyTo: ActorRef[Confirmation]) extends Command
+
+  /**
+   * A command to get the current state of the shopping cart.
+   */
+  final case class Get(replyTo: ActorRef[Summary]) extends Command
+
+  final case class Summary(items: Map[String, Int], checkedOut: Boolean) extends CborSerializable
+
+  sealed trait Confirmation extends CborSerializable
+
+  final case class Accepted(summary: Summary) extends Confirmation
+
+  final case class Rejected(reason: String) extends Confirmation
+
+  /**
+   * This interface defines all the events that the ShoppingCart supports.
+   */
+  sealed trait Event extends CborSerializable {
+    def cartId: String
+  }
+
+  final case class ItemAdded(cartId: String, itemId: String, quantity: Int) extends Event
+
+  final case class ItemRemoved(cartId: String, itemId: String) extends Event
+
+  final case class ItemQuantityAdjusted(cartId: String, itemId: String, newQuantity: Int) extends Event
+
+  final case class CheckedOut(cartId: String, eventTime: Instant) extends Event
+
+  val EntityKey: EntityTypeKey[Command] = EntityTypeKey[Command]("ShoppingCart")
+
+  def init(system: ActorSystem[_]): Unit = {
+    val eventProcessorSettings = EventProcessorSettings(system)
+    ClusterSharding(system).init(Entity(EntityKey) { entityContext =>
+      val n = math.abs(entityContext.entityId.hashCode % eventProcessorSettings.parallelism)
+      val eventProcessorTag = eventProcessorSettings.tagPrefix + "-" + n
+      ShoppingCart(entityContext.entityId, Set(eventProcessorTag))
+    }.withRole("write-model"))
+  }
```

Comment:
Add snapshotting?

Reply:
Thanks, I'll include that.

Maybe I'll add backoff supervision also. That is not very important when using sharding since it will be started again on next message, but can be a good way to not restart it too early and thereby overloading the db.